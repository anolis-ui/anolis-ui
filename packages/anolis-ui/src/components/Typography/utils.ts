const separateObjValuesMemo = new WeakMap();

// Separate properties with objects (responsive values) as value
export const separateObjValues = (x?: Record<string, any>) => {
  if (x === undefined) return undefined;

  if (separateObjValuesMemo.has(x)) {
    return separateObjValuesMemo.get(x);
  }

  const e = Object.entries(x);

  const a = [];
  const b = [];

  for (let i = 0; i < e.length; i++) {
    if (e[i][1] && typeof e[i][1] === "object") {
      a.push(e[i]);
    } else {
      b.push(e[i]);
    }
  }

  const result = [a, b] as const;
  separateObjValuesMemo.set(x, result);
  return result;
};

const groupByBreakpointMemo = new WeakMap();

// Transform from `width: { sm: 2 }` to `sm: { with: 2 }
export const groupByBreakpoint = (x?: [string, Record<string, any>][]) => {
  if (x === undefined) return undefined;

  if (groupByBreakpointMemo.has(x)) {
    return groupByBreakpointMemo.get(x);
  }

  const result: any = {};

  for (let i = 0; i < x.length; i++) {
    const propKey = x[i][0];
    const keys = Object.keys(x[i][1]);

    for (let ii = 0; ii < keys.length; ii++) {
      const breakpointKey = keys[ii];
      result[breakpointKey] = result[breakpointKey] ?? {};
      result[breakpointKey][propKey] = x[i][1][breakpointKey];
    }
  }

  groupByBreakpointMemo.set(x, result);
  return result;
};

// wrap strings (media queries) in an array with css selector
export const wrapObjsWithSelector = (selector: string, arr: (string | {} | null | undefined)[]) => {
  const result = [];

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === null || arr[i] === undefined) continue;

    if (typeof arr[i] !== "string") {
      result.push(`${selector} {`, arr[i], "}");
    } else {
      result.push(arr[i]);
    }
  }
  return result;
};
