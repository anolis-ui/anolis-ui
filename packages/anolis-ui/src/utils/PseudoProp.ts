import { PropsOf } from "@emotion/react";
import { x } from "@xstyled/emotion";

export type PseudoProp<C extends keyof JSX.IntrinsicElements = "div", P extends object = {}> = PropsOf<(typeof x)[C]> & P;
