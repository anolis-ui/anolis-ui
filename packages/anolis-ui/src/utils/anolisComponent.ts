import { SystemProps, x } from "@xstyled/emotion";
import { StyledComponent } from "@emotion/styled";
import hoistNonReactStatics from "hoist-non-react-statics";
import { forwardRef, ForwardRefRenderFunction, Ref, RefObject, ReactNode } from "react";
import { SizeVariantProps } from "./theme";

export type AnolisComponent<
  Element extends keyof JSX.IntrinsicElements,
  Props extends object,
  Variant extends string = never,
  Size extends string = never
> =
  StyledComponent<
  SystemProps & SizeVariantProps<Variant, Size> & { as?: React.ElementType } & Props,
  {},
  Omit<JSX.IntrinsicElements[Element], "color">
  >;

export type AnolisComponentProps<
  Element extends keyof JSX.IntrinsicElements,
  Props extends object,
  Variant extends string = never,
  Size extends string = never
> =
  & AnolisBaseProps<Element>
  & SizeVariantProps<Variant, Size>
  & { as?: React.ElementType }
  & Props;

export type AnolisBaseProps<
  E extends keyof JSX.IntrinsicElements,
  V extends string = never,
  S extends string = never
> =
  & SystemProps
  & Omit<JSX.IntrinsicElements[E], "color" | "v" | "s" | "ref">
  & SizeVariantProps<V, S>
  & {
    as?: React.ElementType;
    ref?: Ref<RefOf<E>>;

    children?: ReactNode;
  };

export const anolisComponent = <
  Element extends keyof JSX.IntrinsicElements,
  Props extends AnolisComponentProps<Element, object, string, string>
>(
  tag: Element,
  Component: ForwardRefRenderFunction<RefOf<Element>, Props>
): AnolisComponent<Element, Props, NonNullable<Props["v"]>, NonNullable<Props["s"]>> => {
  const a = forwardRef(hoistNonReactStatics(Component, x[tag]));

  return a as any;
};

export const anolisComp = <T, P = {}>(displayName: string, render: ForwardRefRenderFunction<T, P>) => {
  render.displayName = `anolis(${displayName})`;

  return forwardRef(render);
};

// Way to get from element key strings ("div", "span", "a") their HTML type (HTMLDivElement, HTMLSpanElement, HTMLAnchorElement)
type RefOf<K extends keyof JSX.IntrinsicElements> =
  Exclude<Parameters<Exclude<(JSX.IntrinsicElements[K]["ref"]), undefined | string | null | RefObject<{}>>>[0], null>;
