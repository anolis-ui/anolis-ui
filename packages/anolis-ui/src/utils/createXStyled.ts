/* eslint-disable @typescript-eslint/no-shadow */
import emStyled, { CreateStyled } from "@emotion/styled";
import { css, system } from "@xstyled/emotion";
import { StyleGenerator } from "@xstyled/system";

// xstyled does not expose a way to make custom "styled.*Box". Components made with styled() does not
// support utility props. This file contains simplified functions to create styled() with utility props.
// Original: https://github.com/gregberge/xstyled/blob/main/packages/emotion/src/createStyled.ts

type XCSSFunction = typeof css;

const getCreateStyle = (baseCreateStyle: any, css: XCSSFunction, generator: StyleGenerator) => {
  return (strings: any, ...args: any) => {
    if (Array.isArray(strings)) {
      strings = [...strings, "\n"];
    }
    args = [...args, generator];
    return baseCreateStyle((props: any) =>
      css(strings, ...flattenArgs(args, props))(props)
    );
  };
};

const createShouldForwardProp = (generator: StyleGenerator) => {
  const propSet = new Set(generator.meta.props);
  return (prop: string): boolean =>
    prop !== "as" && !prop.startsWith("$") && !propSet.has(prop);
};

export const createBaseStyled = (css: XCSSFunction, generator: StyleGenerator): CreateStyled => (
  (component: any, options: any) =>
    getCreateStyle(
      emStyled(component, {
        shouldForwardProp: createShouldForwardProp(generator),
        ...options
      }),
      css,
      generator
    )
) as CreateStyled;

export const createXStyled = <TGen extends StyleGenerator>(generator: TGen) => {
  const xstyled = createBaseStyled(css, generator);

  return xstyled;
};

const styleToString = (style: any, props: any): any => {
  if (Array.isArray(style)) {
    return style.reduce((str, style) => str + styleToString(style, props), "");
  }
  if (typeof style === "function") return styleToString(style(props), props);
  return style;
};

const flattenArgs = (arg: any, props: any): any => {
  if (typeof arg === "function") return flattenArgs(arg(props), props);
  if (Array.isArray(arg)) return arg.map((arg) => flattenArgs(arg, props));
  return arg;
};

export const xstyled = createXStyled(system);
