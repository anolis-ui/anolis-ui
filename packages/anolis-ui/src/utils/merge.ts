/**
 * Returns a new object with properties of input objects (properties starting with "_" are merged deeply, right-to-left priority).
 */
export const merge = <T>(...sources: (T | undefined)[]): T => {
  return _merge<T>(...sources);
};

export const _merge = <T>(...sources: (T | undefined)[]): T => {
  const result: any = {};
  const chain: T[] = [];
  const keys = new Set<string>();

  for (const s of sources) {
    if (s === undefined) continue;
    chain.unshift(s);
    for (const k of Object.keys(s)) {
      keys.delete(k); // always add the key at the end
      keys.add(k);
    }
  }

  for (const key of keys) {
    // eslint-disable-next-line @typescript-eslint/no-extra-parens
    const shouldDeepMerge = key.startsWith("_") && chain.every((s: any) => s[key] === undefined || (s[key] && typeof s[key] === "object"));

    const value = shouldDeepMerge
      ? _merge(...reverse(chain.map(s => (s as any)[key])))
      : (chain as any[]).find(s => !isNil(s[key]))?.[key];

    !isNil(value) && (result[key] = value);
  }

  return result;
};

const isNil = (x: any): x is undefined | null => x === null || x === undefined;
const reverse = <T extends any[]>(x: T) => [...x].reverse();
