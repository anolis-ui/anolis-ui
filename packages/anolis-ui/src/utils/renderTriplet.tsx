import { x } from "@xstyled/emotion";
import { ElementType } from "react";

import renderComponent, { Renderable } from "./renderComponent";
import { ElementProps } from "./theme";

export const renderTriplet = <E extends ElementType>(
  $: E | undefined,
  _: ElementProps<E> | undefined,
  children?: Renderable
) => {
  const Component = typeof $! === "string" ? x[$! as never] : $!;
  return children
    ? <Component {..._!}>{renderComponent(children)}</Component>
    : <Component {..._!} />;
};
