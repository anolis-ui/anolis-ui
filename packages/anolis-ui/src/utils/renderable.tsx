import { ComponentType, ReactNode } from "react";
import { isValidElementType } from "react-is";

export type Renderable = ReactNode | ComponentType;

export const renderable = (C: Renderable): ReactNode =>
  typeof C !== "string" && isValidElementType(C)
    ? <C />
    : (C as ReactNode);
