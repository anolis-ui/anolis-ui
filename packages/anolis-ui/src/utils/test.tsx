import AnolisProvider from "components/AnolisProvider";
import { defaultTheme } from "defaultTheme";
import { emptyTheme } from "theme";
import { FC, ReactElement } from "react";
import { Queries, queries, render, RenderOptions } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import renderer from "react-test-renderer";

expect.extend(toHaveNoViolations);

const Providers: FC = ({ children }) => {
  return (
    <AnolisProvider xstyledTheme={defaultTheme} theme={emptyTheme}>
      {children}
    </AnolisProvider>
  );
};

export const customRender = <
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement
>(ui: ReactElement, options?: RenderOptions<Q, Container>) => render(ui, { wrapper: Providers, ...options });

export const testA11y = async (ui: ReactElement) => {
  const { container } = customRender(ui);

  expect(await axe(container)).toHaveNoViolations();
};

export const testSnapshotMatch = (ui: ReactElement) => {
  const tree = renderer.create(
    <AnolisProvider xstyledTheme={defaultTheme} theme={emptyTheme}>
      {ui}
    </AnolisProvider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
};

export * from "@testing-library/react";

export { customRender as render };
