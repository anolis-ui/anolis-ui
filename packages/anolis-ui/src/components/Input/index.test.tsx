import { render, testA11y, testSnapshotMatch } from "utils/test";

import { Input } from ".";

test("should render", () => {
  render(
    <Input />
  );
});

test("should conform a11y", async () => {
  await testA11y(
    <label>
      Username
      <Input />
    </label>
  );
});

test("should match snapshot", () => {
  testSnapshotMatch(
    <Input />
  );
});
