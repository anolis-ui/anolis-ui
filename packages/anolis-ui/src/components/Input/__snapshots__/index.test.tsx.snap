// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  -webkit-transition: border 300ms,background 300ms;
  transition: border 300ms,background 300ms;
  color: #808296;
  font-size: 1rem;
  line-height: 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  border: 1px solid;
  border-color: #b9b9c6;
  padding-right: 1rem;
  padding-left: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.emotion-0:focus-within {
  color: #505162;
  border-color: #808296;
}

.emotion-0:hover {
  color: #67687d;
  border-color: #9c9eae;
}

.emotion-0:user-invalid,
[data-a-invalid=true] .emotion-0,
[data-a-invalid=true].emotion-0 {
  color: #dc2626;
  border-color: #dc2626;
}

.emotion-0 input,
.emotion-0 textarea {
  background: transparent;
  border: none;
  -webkit-appearance: inherit;
  -webkit-transition: inherit;
  transition: inherit;
  resize: none;
  width: 100%;
  color: inherit;
}

.emotion-0 input:focus,
.emotion-0 textarea:focus {
  border: none;
  outline: none;
}

.emotion-0 input:-webkit-autofill,
.emotion-0 textarea:-webkit-autofill,
.emotion-0 input:-webkit-autofill:hover,
.emotion-0 textarea:-webkit-autofill:hover,
.emotion-0 input:-webkit-autofill:focus,
.emotion-0 textarea:-webkit-autofill:focus,
.emotion-0 input:-webkit-autofill:active,
.emotion-0 textarea:-webkit-autofill:active {
  -webkit-box-shadow: 0 0 0 3rem white inset!important;
}

<div
  className="emotion-0"
  onClick={[Function]}
>
  <input
    className="emotion-1"
  />
</div>
`;
